api/v1alpha1/api_key_auth_types.go

```diff
@@ -0,0 +1,59 @@
+// Copyright Envoy Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package v1alpha1
+
+import (
+	gwapiv1 "sigs.k8s.io/gateway-api/apis/v1"
+)
+
+const APIKeysSecretKey = "credentials"
+
+// APIKeyAuth defines the configuration for the API Key Authentication.
+type APIKeyAuth struct {
+	// Credentials is the Kubernetes secret which contains the API keys.
+	// This is an Opaque secret.
+	// Each API key is stored in the key representing the client id,
+	// which can be used in AllowedClients to authorize the client in a simple way.
+	Credentials gwapiv1.SecretObjectReference `json:"credentials"`
+
+	// KeySources is where to fetch the key from the coming request.
+	// The value from the first source that has a key will be used.
+	KeySources []*KeySource `json:"keySources"`
```

Comment:
I like OpenAPI's definiton which uses `In`
https://swagger.io/docs/specification/v3_0/authentication/api-keys/
```
in:
  type: Header
  name: "x-api-id"
```

another existing design pattern

the `jwt` field uses `extractFrom` https://gateway.envoyproxy.io/docs/api/extension_types/#jwtprovider

Reply:
Well, generally speaking, I do prefer the current form in this kind of case at API design, especially for better future extensibility; it allows us to add a different structure later, e.g., let's say envoy supports a new key source fetching key from body (which Kong already supports), it might be like:

```go
type KeySource struct {
	Header *string 
	Query *string 
	Cookie *string

	Body *BodyKeySource
}

type BodyKeySource struct {
	BodyType string // yaml or json.
	Field string // where we get API Key from.
}
```

Also, `extractFrom` actually looks like very similar to `KeySource` here, are you suggesting renaming it from KeySources to ExtractFrom?

Reply:
yeah suggesting we pick a name thats already used in the API / or industry and also try and use a union discriminator i.e.
```
extractFrom:
  type: Header
  header: <val> 
```

e.g. https://gateway.envoyproxy.io/docs/api/extension_types/#ratelimitspec
https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/1027-api-unions/README.md

Reply:
Renamed to `extractFrom`.